src/java.base/share/classes/jdk/internal/foreign/LayoutRecordMapper.java

```diff
@@ -477,24 +448,35 @@ static int dimensionOf(Class<?> arrayClass) {
 
     // Wrapper to create an array of Records
 
-    @SuppressWarnings("unchecked")
     static <R> R[] toArray(MemorySegment segment,
                            GroupLayout elementLayout,
                            long offset,
                            long count,
-                           Class<?> type,
+                           Class<R> type,
                            MethodHandle mapper) {
 
-        return slice(segment, elementLayout, offset, count)
-                .elements(elementLayout)
-                .map(s -> {
-                    try {
-                        return mapper.invokeExact(s);
-                    } catch (Throwable t) {
-                        throw new IllegalArgumentException(t);
-                    }
-                })
-                .toArray(l -> (R[]) Array.newInstance(type, l));
+        var slice = slice(segment, elementLayout, offset, count);
+        return toArray(slice, elementLayout, count, type, mapper);
+    }
+
+    @SuppressWarnings("unchecked")
+    static <R> R[] toArray(MemorySegment segment,
+                           GroupLayout elementLayout,
+                           long count,
+                           Class<R> type,
+                           MethodHandle mapper) {
+
+        R[] result = (R[]) Array.newInstance(type, Math.toIntExact(count));
+        Spliterator<MemorySegment> spliterator = segment.spliterator(elementLayout);
```

Comment:
Using MemorySegment::elements allows you to avoid interacting with the spliterator directly (which I don't think you need here?)

Reply:
This was an attempt to optimize but I agree we should use a Stream instead.